python.d/linux_power_supply.chart.py

```diff
@@ -0,0 +1,156 @@
+        lines.append([attr_min_design, attr_min_design, 'absolute', 1, PRECISION])
+        attrlist.append(attr_max_design)
+        attrlist.append(attr_min_design)
+    elif _get_sysfs_value(os.path.join(syspath, attr_min_design)) is not None:
+        lines.append([attr_min_design, attr_min_design, 'absolute', 1, PRECISION])
+        attrlist.append(attr_min_design)
+    return {name: {'options': options, 'lines': lines}}, attrlist
+
+def _get_charge_chart(syspath):
+    # Charge is measured in microamphours.  We track up to five
+    # attributes.
+    return _get_generic_chart(syspath, 'charge', 'µAh', 'full', 'empty')
+
+def _get_energy_chart(syspath):
+    # Energy is measured in microwatthours.  We track up to five
+    # attributes.
+    return _get_generic_chart(syspath, 'energy', 'µWh', 'full', 'empty')
+
+def _get_voltage_chart(syspath):
+    # Voltage is measured in microvolts. We track up to five attributes.
+    return _get_generic_chart(syspath, 'voltage', 'µV', 'min', 'max')
+
+# This is a list of functions for generating charts.  Used below to save
+# a bit of code (and to make it a bit easier to add new charts).
+GET_CHART = {
+    'capacity': _get_capacity_chart,
+    'charge': _get_charge_chart,
+    'energy': _get_energy_chart,
+    'voltage': _get_voltage_chart
+}
+
+# This opens the specified file and returns the value in it or None if
+# the file doesn't exist.
+def _get_sysfs_value(filepath):
+    try:
+        with open(filepath, 'r') as datasource:
+            return int(datasource.read())
+    except (OSError, IOError):
+        return None
+
+# Certain attributes (*_empty and *_empty_design) are implicitly zero
+# if not provided by the driver.  This function is used to fetch those
+# attributes.
+def _get_sysfs_value_or_zero(filepath):
+    result = _get_sysfs_value(filepath)
+    if result is None:
+        return 0
+    return result
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.definitions = dict()
+        self.order = list()
+        self.attrlist = list()
+        self.supply = self.configuration.get('supply', None)
+        if self.supply is not None:
+            self.syspath = '/sys/class/power_supply/{0}'.format(self.supply)
+        self.types = self.configuration.get('charts', 'energy charge capacity voltage').split()
+
+    def check(self):
+        if self.supply is None:
+            self.error('No power supply specified for monitoring.')
+            return False
+        if not self.types:
+            self.error('No attributes requested for monitoring.')
+            return False
+        if not os.access(self.syspath, os.R_OK):
+            self.error('Unable to access {0}'.format(self.syspath))
+            return False
+        chartset = set(GET_CHART.keys()).intersection(set(self.types))
```

Comment:
not a big deal but `.keys()` call is not needed

Reply:
please move the chart creating to a separate function

Reply:
Said function would just get called by `check()` anyway, because `check()` succeeding is dependent both on the set being created in this line being non-empty, and on the actual set of charts that can be created out of those requested being non-empty.  In essence, generating the charts here is a side effect of what needs to be done to make sure we can collect data.

Reply:
> Said function would just get called by check() anyway

I know, it's ok